src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -787,8 +787,452 @@ namespace Microsoft.FSharp.Core
+                | false -> raise (Exception "no compilation mapping attribute")
+                | true -> x
+
+            let tryFindCompilationMappingAttributeFromType (typ:Type, res:byref<SourceConstructFlags*int*int>) : bool =
+                let assem = typ.Assembly
+                if (not (obj.ReferenceEquals(assem, null))) && assem.ReflectionOnly then 
+                    tryFindCompilationMappingAttributeFromData (typ.GetCustomAttributesData(), &res)
+                else
+                    tryFindCompilationMappingAttribute (typ.GetCustomAttributes (typeof<CompilationMappingAttribute>,false), &res)
+
+            let tryFindCompilationMappingAttributeFromMemberInfo (info:MemberInfo, res:byref<SourceConstructFlags*int*int>) : bool =
+                let assem = info.DeclaringType.Assembly
+                if (not (obj.ReferenceEquals (assem, null))) && assem.ReflectionOnly then 
+                    tryFindCompilationMappingAttributeFromData (info.GetCustomAttributesData(), &res)
+                else
+                    tryFindCompilationMappingAttribute (info.GetCustomAttributes (typeof<CompilationMappingAttribute>,false), &res)
+
+            let findCompilationMappingAttributeFromMemberInfo (info:MemberInfo) =    
+                let assem = info.DeclaringType.Assembly
+                if (not (obj.ReferenceEquals (assem, null))) && assem.ReflectionOnly then 
+                    findCompilationMappingAttributeFromData (info.GetCustomAttributesData())
+                else
+                    findCompilationMappingAttribute (info.GetCustomAttributes (typeof<CompilationMappingAttribute>,false))
+
+            let tryFindSourceConstructFlagsOfType (typ:Type, res:byref<SourceConstructFlags>) : bool =
+              let mutable x = unsafeDefault<_>
+              if tryFindCompilationMappingAttributeFromType (typ, &x) then
+                let flags,_n,_vn = x
+                res <- flags
+                true
+              else
+                false
+
+            let isKnownType (typ:Type, bindingFlags:BindingFlags, knownType:SourceConstructFlags) =
+                let mutable flags = unsafeDefault<_>
+                match tryFindSourceConstructFlagsOfType (typ, &flags) with 
+                | false -> false 
+                | true ->
+                  (flagsContains flags SourceConstructFlags.KindMask knownType) &&
+                  // We see private representations only if BindingFlags.NonPublic is set
+                  (if flagsIsSet flags SourceConstructFlags.NonPublicRepresentation then 
+                      flagsIsSet bindingFlags BindingFlags.NonPublic
+                   else 
+                      true)
+
+            let isRecordType (typ:Type, bindingFlags:BindingFlags) = isKnownType (typ, bindingFlags, SourceConstructFlags.RecordType)
+            let isObjectType (typ:Type, bindingFlags:BindingFlags) = isKnownType (typ, bindingFlags, SourceConstructFlags.ObjectType)
```

Comment:
To make this clearer, this should be `isFSharpObjectType`. Same with the others: `isFSharpRecordType`, `isFSharpUnionType`.